kernel/aster-nix/src/fs/ext2/inode.rs

```diff
@@ -617,45 +634,39 @@ impl Inner {
 impl Inner {
     pub fn new(desc: Dirty<InodeDesc>, weak_self: Weak<Inode>, fs: Weak<Ext2>) -> Self {
         let num_page_bytes = desc.num_page_bytes();
-        let inode_impl = InodeImpl::new(desc, weak_self, fs);
+        let inode_impl = InodeImpl::new(desc, weak_self, fs.clone());
         Self {
             page_cache: PageCache::with_capacity(num_page_bytes, Arc::downgrade(&inode_impl) as _)
                 .unwrap(),
             inode_impl,
+            fs,
         }
     }
 
+    fn fs(&self) -> Arc<Ext2> {
+        self.fs.upgrade().unwrap()
+    }
+
     pub fn resize(&mut self, new_size: usize) -> Result<()> {
         self.inode_impl.resize(new_size)?;
         self.page_cache.pages().resize(new_size)?;
         Ok(())
     }
 
-    pub fn read_at(&self, offset: usize, buf: &mut [u8]) -> Result<usize> {
-        let (offset, read_len) = {
-            let file_size = self.inode_impl.file_size();
-            let start = file_size.min(offset);
-            let end = file_size.min(offset + buf.len());
-            (start, end - start)
-        };
+    pub fn read_at(&self, offset: usize, buf: &mut [u8]) -> Result<()> {
+        self.page_cache.pages().read_bytes(offset, buf)?;
 
-        self.page_cache
-            .pages()
-            .read_bytes(offset, &mut buf[..read_len])?;
-        Ok(read_len)
+        self.inode_impl
+            .set_atime(self.fs().time_provider().current());
```

Comment:
Calling `set_atime` in `read_at` makes it impossible to support `O_NOATIME` flag, which may be set when open a file or using `fcntl`.